### PR TITLE
chore(logger): streamline logging

### PR DESCRIPTION
**Pull request by SimonSchick**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Replaces `debug` package with `debuglog` native and acts as a foundation for https://github.com/sequelize/sequelize/pull/10284#discussion_r269296280

cc @javiertury 

~~#Fixes 10661~~

### CODE REVIEW

**Comment by eseliger** [comment]
Looking good so far, but the build is failing

**Comment by SimonSchick** [comment]
hnnng

**Comment by SimonSchick** [comment]
Yes, see https://nodejs.org/api/util.html#util_util_debuglog_section

**Comment by SimonSchick** [comment]
I actually went ahead and streamlined all query logging into a single method `sequelize.logQuery` that handles both debug and regular logging mechanisms.
I also managed to purge a good amount of duplication while doing so.

Please note that it's relatively _easy_ to bring back the `debug` package if that's highly desired.

**Comment by eseliger** [comment]
to me this looks very clean and like the way we want to be headed to. So if you guys feel like this is a breaking change I'd still be up for pushing this for a v6 release instead 🤔

**Comment by SimonSchick** [comment]
I don't think this is breaking btw, in no place sequelize states that it uses `debug` or that you can use `DEBUG=sequelize` as env do enable debugging, it's not part of any public interfaces either.

**Comment by eseliger** [comment]
I'm totally up for merging this, but the build is broken again 😔

**Comment by SimonSchick** [comment]
Heck.

File: lib/dialects/postgres/query.js [hunk 1]
```diff
@@ -43,28 +43,18 @@ class Query extends AbstractQuery {
   }
 
   run(sql, parameters) {
-    const connection = this.connection;
-    this.sql = sql;
+    const { connection } = this;
 
     if (!_.isEmpty(this.options.searchPath)) {
-      this.sql = this.sequelize.getQueryInterface().QueryGenerator.setSearchPath(this.options.searchPath) + sql;
+      sql = this.sequelize.getQueryInterface().QueryGenerator.setSearchPath(this.options.searchPath) + sql;
```
Comment (eseliger) [inline lib/dialects/postgres/query.js]: hmm I think the idea was before to not change the input string by using the "copy" above. I don't know if there is any assumptions made it is immutable or not 🤔 but if the tests are passing, should be ok 
Also, same with the above

**Comment by sushantdhiman** [comment]
This is a breaking change and can't be merged. Even if docs are lacking that we support `debug` I am sure many are using this feature since we added support for `debug`. Also the environment variable for logs will be changed from `DEBUG` to `NODE_DEBUG`

**Comment by SimonSchick** [comment]
Imo the only _breaking_ change is the change in any *public* documentation or otherwise explicitly exposed format.

How would anyone know you can use `debug` if it's not in the docs?
They'd have to inspect the actual code to figure that out, and add that point 99% of developers would probably just add `console.log` at that point or use the `logging` option instead.

I have reverted it but please consider documenting the things you want developers to use/know about.

**Comment by SimonSchick** [comment]
Boiled it down a little, the `DebugContext` class was a remnant of me trying make logging more lazy, ended up not using it.

**Comment by SimonSchick** [comment]
@sushantdhiman snug in a fix for #10661

File: lib/utils/logger.js [hunk 2]
```diff
@@ -11,38 +11,34 @@
 const debug = require('debug');
```
Comment (sushantdhiman) [inline lib/utils/logger.js]: Not too happy with changes in this class.

1) You are caching `this.logger` on `Logger` class (that may confusing as `this` is used in static method). Make a const declaration after class with is an instance of class as `logger = new Logger()`.

2) We can probably get rid of both old and new `getLogger` method, just export instance `logger` of `Logger` class created in step (1), require and use that everywhere

File: lib/utils/logger.js [hunk 3]
```diff
@@ -11,38 +11,34 @@
...
   }
 
   warn(message) {
     // eslint-disable-next-line no-console
     console.warn(`(${this.config.context}) Warning: ${message}`);
   }
 
-  debugContext(childContext) {
-    if (!childContext) {
-      throw new Error('No context supplied to debug');
```
Comment (sushantdhiman) [inline lib/utils/logger.js]: Can probably let it slide, but this empty check may prevent logging to `sequelize:undefined` etc

File: lib/dialects/abstract/query.js [hunk 4]
```diff
@@ -320,6 +320,41 @@ class AbstractQuery {
...
+    };
+  }
+
+  /**
+   * Gets a debug context for sub classes.
+   *
+   * @param {string} dialect
+   * @private
+   */
+  static makeDebugContext(dialect) {
```
Comment (sushantdhiman) [inline lib/dialects/abstract/query.js]: Get rid of this method, let dialect query class create debug context by old `logger.debugContext(...)` method

File: lib/dialects/mariadb/query.js [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
 const ER_DUP_ENTRY = 1062;
 const ER_ROW_IS_REFERENCED = 1451;
 const ER_NO_REFERENCED_ROW = 1452;
 
+const debug = AbstractQuery.makeDebugContext('mariadb');
+
 class Query extends AbstractQuery {
   constructor(connection, sequelize, options) {
-    super(connection, sequelize, Object.assign({ showWarnings: false }, options));
+    super(connection, sequelize, debug, Object.assign({ showWarnings: false }, options));
```
Comment (sushantdhiman) [inline lib/dialects/mariadb/query.js]: Why debug is passed to super.constructor, `AbstractQuery` is not using debug in constructor, third argument should be `options`.

Comment (SimonSchick) [reply lib/dialects/mariadb/query.js]: Uh oh.

File: lib/utils/logger.js [hunk 6]
```diff
@@ -11,38 +11,28 @@
 const debug = require('debug');
 
 class Logger {
+  static warn(message) {
```
Comment (sushantdhiman) [inline lib/utils/logger.js]: static `warn` can be removed, just use `warn` from instance exported

File: lib/dialects/abstract/query.js (hunk 4 above)
Comment (sushantdhiman) [reply lib/dialects/abstract/query.js]: Resolved? `AbstractQuery` shouldn't be responsible for this, let's got with simple approach?

Comment (SimonSchick) [reply lib/dialects/abstract/query.js]: Oh, final push didn't go through 😅

**Pull request merged**